pkg/operator/operator_starter.go

```diff
@@ -0,0 +1,306 @@
+package operator
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/api/sharedresource"
+	"github.com/openshift/cluster-storage-operator/pkg/csoclients"
+	"github.com/openshift/cluster-storage-operator/pkg/operator/configobservation/configobservercontroller"
+	"github.com/openshift/cluster-storage-operator/pkg/operator/csidriveroperator"
+	"github.com/openshift/cluster-storage-operator/pkg/operator/csidriveroperator/csioperatorclient"
+	"github.com/openshift/cluster-storage-operator/pkg/operator/defaultstorageclass"
+	"github.com/openshift/cluster-storage-operator/pkg/operator/vsphereproblemdetector"
+	"github.com/openshift/cluster-storage-operator/pkg/operatorclient"
+	"github.com/openshift/library-go/pkg/controller/controllercmd"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/configobserver/featuregates"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/loglevel"
+	"github.com/openshift/library-go/pkg/operator/management"
+	"github.com/openshift/library-go/pkg/operator/managementstatecontroller"
+	"github.com/openshift/library-go/pkg/operator/status"
+	rbacv1 "k8s.io/api/rbac/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/klog/v2"
+)
+
+type OperatorStarter interface {
+	StartOperator(ctx context.Context) error
+	initClient(ctx context.Context) error
+}
+
+type commonStarter struct {
+	controllerConfig *controllercmd.ControllerContext
+
+	eventRecorder events.Recorder
+	versionGetter status.VersionGetter
+	featureGates  featuregates.FeatureGate
+
+	commonClients *csoclients.Clients
+	// array of controllers that needs to be started.
+	controllers []factory.Controller
+}
+
+func (csr *commonStarter) initClient(ctx context.Context) error {
+	clients, err := csoclients.NewClients(csr.controllerConfig, resync)
+	if err != nil {
+		return err
+	}
+	csr.commonClients = clients
+	csr.eventRecorder = csr.controllerConfig.EventRecorder
+	return nil
+}
+
+func (csr *commonStarter) getFeatureGate(ctx context.Context, recorder events.Recorder) error {
+	desiredVersion := status.VersionForOperatorFromEnv()
+	missingVersion := "0.0.1-snapshot"
+
+	// By default, this will exit(0) the process if the featuregates ever change to a different set of values.
+	featureGateAccessor := featuregates.NewFeatureGateAccess(
+		desiredVersion, missingVersion,
+		csr.commonClients.ConfigInformers.Config().V1().ClusterVersions(), csr.commonClients.ConfigInformers.Config().V1().FeatureGates(),
```

Comment:
This will read the feature gates from the mgmt cluster in HyperShift. It should use guest client + informer (passed as arguments to `getFeatureGate`?)